presto-hive/src/main/java/io/prestosql/plugin/hive/util/ThrottledAsyncQueue.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.util;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.log.Logger;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+
+/**
+ * An asynchronous queue that limits the rate at which batches will be made available as well as the number
+ * of elements they will contain.
+ *
+ * @param <T>           The type of elements accepted by the queue.
+ */
+@ThreadSafe
+public class ThrottledAsyncQueue<T>
+        extends AsyncQueue<T>
```

Comment:
Do we need to implement `AsyncQueue`?
The `ThrottledAsyncQueue` class look complex and the problem at hand looks simple ("simplish": simple + parallelizm and futures).

Did you look whether it is possible it achieve same thing with Guava `RateLimiter` ?

(i didn't review this class yet)

Reply:
The ThrottledAsyncQueue is an "extension" of the AsyncQueue that only throttles the polling of the batches, the rest of the logic is inherited from `AsyncQueue`.

Thanks for pointing the `RateLimiter`, I wasn't aware of it. I pushed a commit to use it and it simplifies greatly the cost, however it changes the logic slightly:
In my previous attempt, the `borrowBatchAsync()` method would favor immediately (and synchronously) returning a small batch before throttling if possible. With `RateLimiter`, the behavior is to prefer waiting in favor of returning larger batches later.
For instance, with `hive.max-splits-per-second = 10`, if we have already de-queued a first batch of 8 elements and there are still 4 elements in the queue, my first implementation (without `RateLimiter`) would immediately return a batch of size 2 (as an immediate future). The `RateLimiter`, however, would block for a short time and return a batch of size 4.

I don't think it makes a major difference in the real world, and can make the code more readable at the cost of having less batches returned synchronously.